### PR TITLE
Flame Graph: Clearer x axis

### PR DESCRIPTION
**Pull request by joey-grafana**
**What this PR does / why we need it**:

This PR makes it clearer as to what the flame graphs x-axis represents by using units and ticks that update when a node is clicked on.

### CODE REVIEW

File: public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraphTooltip.tsx [hunk 1]
```diff
@@ -45,13 +45,28 @@ const FlameGraphTooltip = ({ tooltipRef, tooltipData, showTooltip }: Props) => {
...
+    percentValue: percent,
+    unitTitle: titles.unitTitle,
+    unitValue,
+    samples: samples.toLocaleString(),
+  };
+};
+
+const getTitles = (field: Field) => {
+  let percentTitle = '';
+  let unitTitle = '';
```
Comment (joey-grafana) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraphTooltip.tsx:69]: ok so not exactly relevant to this PR. Small tidy up :D

**Comment by joey-grafana** [comment]
Going to update this to show an actual x-axis with a range of e.g. 0 - 20 GB that updates when a block is clicked

**Comment by leeoniya** [comment]
i'll throw in my $0.02 here and say that adding an x axis for this purpose is not a good approach.

the request in the original issue is to clarify what units is dimension is represented horizontally in the flame graph.

while the unit may differ depending on the flame graph, it will always be cardinal and _not_ ordinal [1]. in some visualizations a y axis does represent cardinal units (quantity of something), such as a bar chart or histogram -- more up means more quantity. however, in a flame chart the width of each cell represents the quantity; the position of the cell does not matter (more to the right does _not_ mean more quantity).

the current implementation has ticks that imply an ordinal x dimension, which is misleading. yes each tick is 2.5hr wide, but t is not an earlier or later 2.5hrs:

image

what we do care about is the total width of the zoomed region (10hrs), we can then use that to infer that something half as wide is 5hrs, regardless of its position along the "axis", e.g. "60" in this CAD drawing:

image

there are sometimes detached distance scales that do not try to look like an axis along the entire length of the map:

image

generally, my preference would be to simply indicate the full width and units of the zoomed range without making it look like an ordinal axis (which it is not!). another option is to append the raw values to label(s) of the focused elements:

image

[1] https://www.mathsisfun.com/numbers/cardinal-ordinal-chart.html

**Comment by aocenas** [comment]
@leeoniya thanks for the comment and yeah I see the point and agree it may not be ideal fit. My intention was to provide something like the map detached distance scales to give you some sort of scaling.

> my preference would be to simply indicate the full width and units of the zoomed range without making it look like an ordinal axis (which it is not!)

How would you imagine that? Should we just have something more textual? Like `Showing 320ms range`?

Or maybe an option would be to have a minimap similar to what we have in trace view?
Screenshot from 2022-10-27 17-07-58

**Comment by joey-grafana** [comment]
A distance scale that resizes according to flame graph width seems like the way to go. It's detached and accurately represents the axis. Also adding the units to the bar label will make resource consumption a lot clearer. Will have an update today.

**Comment by joey-grafana** [comment]
I created another [pr](https://github.com/grafana/grafana/pull/57959) to add the values and units to each block. It's a fairly simple change that shouldn't require much discussion.

I added a scale to this pr;
- is disconnected from the flame graph
- will always be 50% of the width of the flame graph

Screenshot 2022-11-01 at 13 34 12

Screenshot 2022-11-01 at 13 47 38

The top table is pushed down to be level with the flame graph. I'm not overly excited about the design here, but this felt like one of the better options.

**Comment by leeoniya** [comment]
imo the units in the blocks are a sufficient (and better) reference, since they're more adjacent to other blocks you might want to guesstimate.

**Comment by joey-grafana** [comment]
> it doesnt hurt to also state the overall x size in the header text since it has better contrast and larger font size. not sure whether this header text should change with zoom tho.

Sorry not completely clear what you mean by header text. In the header we have the search bar, reset button and select view buttons.

If we put another header above the flame graph instead of the scale, we'll be pushing the top table down, which is ok just not sure about the design. If we put the overall x size just above the flame graph, it's going to be the same as the unit value on the total block below it anyway.

**Comment by leeoniya** [comment]
> If we put another header above the flame graph instead of the scale, we'll be pushing the top table down, which is ok just not sure about the design. If we put the overall x size just above the flame graph, it's going to be the same as the unit value on the total block below it anyway.

agreed, it would add more ui clutter & duplication of info. i was just remembering (your?) other PR that added some title text above the flame graph :)

**Pull request closed without merging**